File: leetcode3/염혜정/1329. Sort the Matrix Diagonally.java
class Solution {
    public int[][] diagonalSort(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        Map<Integer, PriorityQueue<Integer>> map = new HashMap<>();

        // 1. 대각선별로 값 모으기 (key = row - col)
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int key = i - j;
                map.computeIfAbsent(key, k -> new PriorityQueue<>()).offer(mat[i][j]);
            }
        }

        // 2. 정렬된 값 다시 채워넣기
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int key = i - j;
                mat[i][j] = map.get(key).poll();
            }
        }

        return mat;
    }
}

File: leetcode3/최원준/1329. Sort the Matrix Diagonally.py
class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        def helper(a, b):
            nums = []
            row = a
            col = b
            while row < n and col < m:
                nums.append(mat[row][col])
                row += 1
                col += 1
            nums.sort()

            row = a
            col = b
            idx = 0
            while row < n and col < m:
                mat[row][col] = nums[idx]
                row += 1
                col += 1
                idx += 1

        n = len(mat)
        m = len(mat[0])

        for top in range(m):
            helper(0, top)           
        
        for left in range(1, n):
            helper(left, 0)

        return mat
            
